From 4d5604c6c89c69ac7b68adab0f200432a0050b3c Mon Sep 17 00:00:00 2001
From: peak3d <pfau@peak3d.de>
Date: Fri, 4 Aug 2017 00:32:41 +0200
Subject: [PATCH 05/16] [HLS] soft reset on read errors on eosegment

---
 src/tsDemuxer.cpp | 7 ++++---
 src/tsDemuxer.h   | 2 +-
 2 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/src/tsDemuxer.cpp b/src/tsDemuxer.cpp
index 15b8034..76dfacb 100644
--- a/src/tsDemuxer.cpp
+++ b/src/tsDemuxer.cpp
@@ -358,12 +358,13 @@ uint64_t AVContext::Shift()
   return av_pos;
 }
 
-void AVContext::GoPosition(uint64_t pos)
+void AVContext::GoPosition(uint64_t pos, bool rp)
 {
   av_pos = pos;
   Reset();
-  for (std::map<uint16_t, Packet>::iterator it = this->packets.begin(); it != this->packets.end(); ++it)
-    it->second.Reset();
+  if(rp)
+    for (std::map<uint16_t, Packet>::iterator it = this->packets.begin(); it != this->packets.end(); ++it)
+      it->second.Reset();
 }
 
 uint64_t AVContext::GetPosition() const
diff --git a/src/tsDemuxer.h b/src/tsDemuxer.h
index cc70766..a5ab500 100644
--- a/src/tsDemuxer.h
+++ b/src/tsDemuxer.h
@@ -80,7 +80,7 @@ namespace TSDemux
     int TSResync();
     uint64_t GoNext();
     uint64_t Shift();
-    void GoPosition(uint64_t pos);
+    void GoPosition(uint64_t pos, bool rp);
     uint64_t GetPosition() const;
     uint64_t GetNextPosition() const;
     int ProcessTSPacket();
-- 
2.17.1.windows.2

